Framework/Geometry/src/Instrument/InfoComponentVisitor.cpp

```diff
@@ -27,73 +29,138 @@ makeDetIdToIndexMap(const std::vector<detid_t> &detIds) {
-    child->registerContents(*this);
+    children[i] = assemblyChildren[i]->registerContents(*this);
   }
   const size_t detectorStop = m_assemblySortedDetectorIndices->size();
+  const size_t componentIndex = m_componentIds->size();
+  m_assemblySortedComponentIndices->push_back(componentIndex);
+  m_parentComponentIndices->push_back(componentIndex);
+  const size_t componentStop = m_assemblySortedComponentIndices->size();
 
-  m_ranges->emplace_back(std::make_pair(detectorStart, detectorStop));
+  m_detectorRanges->emplace_back(std::make_pair(detectorStart, detectorStop));
+  m_componentRanges->emplace_back(
+      std::make_pair(componentStart, componentStop));
 
   // Record the ID -> index mapping
-  (*m_componentIdToIndexMap)[assembly.getComponentID()] =
-      m_componentIds->size();
-  // For any non-detector we extend the m_componetIds from the back
+  (*m_componentIdToIndexMap)[assembly.getComponentID()] = componentIndex;
+  // For any non-detector we extend the m_componentIds from the back
   m_componentIds->emplace_back(assembly.getComponentID());
+  m_positions->emplace_back(Kernel::toVector3d(assembly.getPos()));
+  m_rotations->emplace_back(Kernel::toQuaterniond(assembly.getRotation()));
+  clearPositionAndRotationParameters(m_pmap, assembly);
+  // Now that we know what the index of the parent is we can apply it to the
+  // children
+  for (const auto &child : children) {
+    (*m_parentComponentIndices)[child] = componentIndex;
+  }
+  markAsSourceOrSample(assembly.getComponentID(), componentIndex);
+  return componentIndex;
 }
 
 /**
  * @brief InfoComponentVisitor::registerGenericComponent
  * @param component : IComponent being visited
+ * @return Component index of this component
  */
-void InfoComponentVisitor::registerGenericComponent(
-    const IComponent &component) {
+size_t
+InfoComponentVisitor::registerGenericComponent(const IComponent &component) {
   /*
    * For a generic leaf component we extend the component ids list, but
    * the detector indexes entries will of course be empty
    */
-  m_ranges->emplace_back(std::make_pair(0, 0)); // Represents an empty range
+  m_detectorRanges->emplace_back(
+      std::make_pair(0, 0)); // Represents an empty range
   // Record the ID -> index mapping
-  (*m_componentIdToIndexMap)[component.getComponentID()] =
-      m_componentIds->size();
+  const size_t componentIndex = m_componentIds->size();
+  (*m_componentIdToIndexMap)[component.getComponentID()] = componentIndex;
   m_componentIds->emplace_back(component.getComponentID());
+  m_positions->emplace_back(Kernel::toVector3d(component.getPos()));
+  m_rotations->emplace_back(Kernel::toQuaterniond(component.getRotation()));
+  const size_t componentStart = m_assemblySortedComponentIndices->size();
+  m_componentRanges->emplace_back(
+      std::make_pair(componentStart, componentStart + 1));
+  m_assemblySortedComponentIndices->push_back(componentIndex);
+  m_parentComponentIndices->push_back(componentIndex);
```

Comment:
Yes you need to extend this vector, but as you are assigning the wrong index here (it is not known yet) maybe a comment would help.